### DISCUSSION TITLE
Upsert functionality

### DISCUSSION THREAD

**Discussion by rjmackay** [sensedeep/dynamodb-onetable#149] [Ideas]
I have several usecases where I'm using `update()` or `create()` to implement an Upsert. However, it's awkward to do, and has a few nuances that are easy to mess up (ie. update not setting `_type` because I disabled the `exists` check).
It would be great to have a built in upsert function, or maybe 2 since you can do it as 'create or overwrite' or 'create or update'

**Comment by mobsense** [comment]
[EDITED] Sounds feasible. However, the intent was that create with exists: null and update with exists: null would do an upsert. So I need to understand what is not working.

Can you list completely the use cases with update/create and exists so I can fully understand the issue. I'd also like to fix all issues with exists that are just bugs. (like not setting _type).

**Comment by rjmackay** [comment]
Here's a couple of examples from our code.

This is a simple create-or-overwrite
```
export const createMarketplaceProfile = async (profile: MarketplaceProfile): Promise<Metrics> => {
  return MarketplaceProfileModel.create({ ...profile }, { exists: null }) // overwrite if exists
}
```

And a create-or-update. In this case I specifically don't want to overwrite properties that are not passed in, so we need an UpdateItem call, not PutItem. I've had to work around a few gotchas: indexes aren't updated unless set explicitly and `_type` isn't set. I'm also passing _type via `set` because otherwise TypeScript complains since the `Route` type doesn't include `_type`
```
export async function saveRoute(route: Route): Promise<Route> {
    const savedRoute = await RouteModel.update(route, {
      set: { _type: 'Route' },
      updateIndexes: true,
      exists: null,
    })
    return savedRoute
}
```

**Reply by rjmackay** [reply]
These aren't too bad by themselves but the full code for saveRoute includes some other where conditions and it's easy to lose which bits are part of the upsert, and which are there for other logic.

**Reply by mobsense** [reply]
Great, thank you that really helps.

We could change update to always set the type and updateIndexes if exists is null. These are currently defaulted to false for performance, but when you set exists to null, you are explicitly saying that a create may be required.

Would that meet the need or is there something extra an `upsert` API would achieve?  That way, the `exists: null` is the single parameter required to drive this behavior?

**Reply by rjmackay** [reply]
I think that could work. However, I worry that that is adding more implicit behaviour, and could potentially be a breaking change for some users.

**Reply by mobsense** [reply]
Understand. But that was the explicit intent of `exists: null`.  You are uncovering what are really bugs I think.

**Reply by rjmackay** [reply]
That makes sense. In which case as long as the docs are clear about when to use exists then I think that make sense.

**Comment by mobsense** [comment]
Fixing the updateIndexes portion in 1.9.1.
